### PR TITLE
Feature/checkpoint-migration

### PR DESCRIPTION
**Pull request by KhoiKy**
* Added `Checkpoint-Migration` function that checkpoints the current state of the database so that it can be re-created.

### CODE REVIEW

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: The first sentence makes it sounds like the main purpose of this function is to push migrations to a database which is actually just a side effect of its main purpose, exporting current state.

Reword to something like:

> The `Checkpoint-Migration` function captures the state of a database after all migrations have been applied. The captured state is exported to a `schema.ps1` file that can be applied with Rivet to re-create that state of the database. Before capturing the current state, a Rivet Push is invoked to ensure all migrations have been applied.

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: The docs say a `schema.ps1` file is created but we're just returning migration objects here?

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: Why are we joining these migration objects with new lines?

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: I think we should pipe this output to `Write-Debug` instead of `Out-Null` in case we ever need it when debugging. In addition we should probably write a debug message just above this that says we're invoking a Rivet push from within the Checkpoint-Migration function.

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: Do we need this try/catch with this error message? I think we can just let `Invoke-Rivet` write the error if it fails.

**Review by decoyjoe — CHANGES_REQUESTED** [review]

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (KhoiKy) [reply Rivet/Functions/Checkpoint-Migration.ps1]: Export-Migration returns the migration as an array list which each line as an array item.

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: So are the objects returned from `Export-Migration` just strings that represent the migration as it would be defined in a .ps1 file?

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: The docs say that the migrations are exported to a `schema.ps1` file, which is not what this is doing. Also a couple of other things:

1. Where should the `schema.ps1` file be placed by default? I think maybe by default gets placed next to the `rivet.json` file. We should probably add an `OutputPath` parameter that allows to customize the path to that file.
2. We should check if the `schema.ps1` file exists first and if it does then fail. Add a `Force` switch that allows overwriting an existing `schema.ps1` file.

**Review by decoyjoe — CHANGES_REQUESTED** [review]

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (KhoiKy) [reply Rivet/Functions/Checkpoint-Migration.ps1]: Yes

**Comment by KhoiKy** [comment]
@decoyjoe

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: We always want to check if this already file exists, not only if an explicit path is given. Change the condition to:

```powershell
if ((Test-Path -Path $OutputPath) -and -not $Force)
```

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: You're assuming the file name is always `schema.ps1` when it may be different if an explicit path was passed in to `$OutputPath`. Change this error message to:

```powershell
Write-Error "Checkpoint output path ""$($OutputPaht)"" already exists. Use the -Force switch to overwrite."
```

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: This is going to place the `schema.ps1` file in the directory where the Rivet module is installed. The `schema.ps1` should default to the same directory as the `rivet.json` file. Is there a way we can resolve that path?

**Review by decoyjoe — CHANGES_REQUESTED** [review]

**Comment by KhoiKy** [comment]
@decoyjoe Once more please.

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1:50]: Add a `return` here after the `Write-Error`.

File: Rivet/Functions/Checkpoint-Migration.ps1
Comment (decoyjoe) [inline Rivet/Functions/Checkpoint-Migration.ps1]: Update the parameter doc to indicate that the default path to the `schema.ps1` file is the same directory as the `rivet.json` file.

File: Rivet/Functions/Invoke-Rivet.ps1
Comment (decoyjoe) [inline Rivet/Functions/Invoke-Rivet.ps1]: Change `$Checkpoint` to `-Checkpoint`

**Review by decoyjoe — DISMISSED** [review]
Looks good but let's wait for a review from @splatteredbits since he wrote up the story.

**Comment by KhoiKy** [comment]
@decoyjoe Sorry once more please. When I force pushed it dismissed the approval :(

**Review by decoyjoe — APPROVED** [review]

**Pull request merged**
